core/store/src/db/recoverydb.rs

```diff
@@ -0,0 +1,222 @@
+}
+
+impl Database for RecoveryDB {
+    /// Returns raw bytes for given `key` ignoring any reference count decoding if any.
+    fn get_raw_bytes(&self, col: DBCol, key: &[u8]) -> std::io::Result<Option<DBSlice<'_>>> {
+        self.cold.get_raw_bytes(col, key)
+    }
+
+    /// Returns value for given `key` forcing a reference count decoding.
+    fn get_with_rc_stripped(&self, col: DBCol, key: &[u8]) -> std::io::Result<Option<DBSlice<'_>>> {
+        self.cold.get_with_rc_stripped(col, key)
+    }
+
+    /// Iterates over all values in a column.
+    fn iter<'a>(&'a self, col: DBCol) -> DBIterator<'a> {
+        self.cold.iter(col)
+    }
+
+    /// Iterates over values in a given column whose key has given prefix.
+    fn iter_prefix<'a>(&'a self, col: DBCol, key_prefix: &'a [u8]) -> DBIterator<'a> {
+        self.cold.iter_prefix(col, key_prefix)
+    }
+
+    /// Iterate over items in given column bypassing reference count decoding if any.
+    fn iter_raw_bytes<'a>(&'a self, col: DBCol) -> DBIterator<'a> {
+        self.cold.iter_raw_bytes(col)
+    }
+
+    /// Iterate over items in given column whose keys are between [lower_bound, upper_bound)
+    fn iter_range<'a>(
+        &'a self,
+        col: DBCol,
+        lower_bound: Option<&[u8]>,
+        upper_bound: Option<&[u8]>,
+    ) -> DBIterator<'a> {
+        self.cold.iter_range(col, lower_bound, upper_bound)
+    }
+
+    /// Atomically applies operations in given transaction.
+    fn write(&self, mut transaction: DBTransaction) -> std::io::Result<()> {
+        self.filter_db_ops(&mut transaction);
+        if !transaction.ops.is_empty() {
+            self.cold.write(transaction)
+        } else {
+            Ok(())
+        }
+    }
+
+    fn compact(&self) -> std::io::Result<()> {
+        self.cold.compact()
+    }
+
+    fn flush(&self) -> std::io::Result<()> {
+        self.cold.flush()
+    }
+
+    fn get_store_statistics(&self) -> Option<crate::StoreStatistics> {
+        self.cold.get_store_statistics()
+    }
+
+    fn create_checkpoint(
+        &self,
+        path: &std::path::Path,
+        columns_to_keep: Option<&[DBCol]>,
+    ) -> anyhow::Result<()> {
+        self.cold.create_checkpoint(path, columns_to_keep)
+    }
+}
+
+impl RecoveryDB {
+    pub fn new(cold: Arc<ColdDB>) -> Self {
+        Self { cold }
+    }
+
+    /// Filters out deletes and other operation which aren't adding new data to the DB.
+    fn filter_db_ops(&self, transaction: &mut DBTransaction) {
+        let mut idx = 0;
+        while idx < transaction.ops.len() {
+            if self.keep_db_op(&mut transaction.ops[idx]) {
+                idx += 1;
+            } else {
+                transaction.ops.swap_remove(idx);
+            }
+        }
+    }
+
+    /// Returns whether the operation should be kept or dropped.
+    fn keep_db_op(&self, op: &mut DBOp) -> bool {
+        let overwrites_same_data = |col: &mut DBCol, key: &mut Vec<u8>, value: &mut Vec<u8>| {
```

Comment:
I'm not sure about this pattern of read-before-write to avoid re-writing the entire DB. It should help in theory, but I don't see any speed up in practice.

Reply:
Well, at least it helps preventing the so-common write stalls.